rfc-0001/README.md

```diff
@@ -0,0 +1,331 @@
+   minute amount of data, a simple update is fine, but if the relation is large, a batched approach is better.
+3. Better implementations may become available (or known at a later time)
+
+We might want to propose using a sort of declarative language that would capture the essence of changes better. Most of the operations can be expressed
+in a form of "morphism" primitive. 
+
+A set of relations, and a set of columns in each, "consumes" them and produces a set of relations 
+with a set of columns (the relations do not need to be new):
+
+```
+[emp(id, email)] => [employee(id, email, name)]
+```
+
+or
+
+```
+[emp(id, email)] => [employee(id, email)]
+[] => [employee(name)]
+```
+
+(the notation is omitting details for brevity)
+
+This can be further lifted into **split/merge** morphism, where one is taking relations
+and splits the columns into other relations. Merge does the reverse. Going one step further,
+some of these can be specialized into **normalize/denormalize** morphisms which capture the essence
+of these common operations even more accurately by identifying their purpose.
+
+So, if we design such a small language and augment it with transformation functions (which will
+now be easier to design a convention for given the morphism structure), we can replace `migration/undo.sql`
+with just one such definition.
+
+However, by itself, it won't be able to enact the migration. For these purposes, `omni_schema` would
+need to be able to do so. The really interesting benefit of such DSL is that we can now enact the migration
+in different ways:
+
+* We can naively translate it to SQL statements. They are going to be of a destructive/interruptive nature,
+  but they will do the job as most migrations do.
+* We can apply rolling updates in schemas, similar to [pgroll](https://pgroll.com/)
+
+### What about data?
+
+Some data in schemas is structural to the application. For example, currencies used in the system or
+types of accounts. We need to be able to migrate those into staging and production, too.
+
+To this end, this proposal recommends observing if any of the tracked relations contain any data and
+capturing them in data files (text/csv/json?)
+
+```
+0194f2ae-5a6f-762a-8ec4-d06d609f14fc/
+  metadata.yaml
+  sources.dat
+  PARENT_ID.diff.yaml
+  migrate.sql 
+  transformations/
+  data/currency.csv
+  data/account_types.csv
+```
+
+In most cases, we just want to ensure that records in those relations make it into migrated schema (and we can
+use this capture as a test).
+
+However, sometimes we need to stipulate that some data must be gone. We can perhaps use some metadata to specify
+if data in some of the tables is meant to be present to the exclusion of any other data.
+
+```
+0194f2ae-5a6f-762a-8ec4-d06d609f14fc/
+  metadata.yaml
+  sources.dat
+  PARENT_ID.diff.yaml
+  migrate.sql 
+  transformations/
+  data/currency.csv
+  data/account_types.csv
+  data/metadata.yaml
+```
+
+***data/metadata.yaml**:
```

Comment:
Do we want to use some implicit value, such as disregarding tables that are not listed? In other words, would it contain an exhaustive list of tables with data?

Reply:
Good question. Implicit is not great (defaults can change). What if we always listed all tables' preferences? Would that be okay? Gives perfect understanding as to where things should stand.